File: solution/0037____Sudoku Solver.go
package main

func solveSudoku(board [][]byte) {
	_0037(board)
}

var row [9][10]byte
var col [9][10]byte
var grid [3][3][10]byte

// perf: 0ms, 84
func _0037(board [][]byte) {
	for i := range row {
		for j := range row[i] {
			row[i][j] = 0
			col[i][j] = 0
		}
	}
	for i := range grid {
		for j := range grid[i] {
			for k := range grid[i][j] {
				grid[i][j][k] = 0
			}
		}
	}

	for i := range board {
		for j := range board[i] {
			if board[i][j] == '.' {
				continue
			} else {
				num := byte(board[i][j] - '0')
				row[i][num] = 1
				col[j][num] = 1
				grid[i/3][j/3][num] = 1
			}
		}
	}
	_0037_helper(board, 0, 0)
}

func _0037_helper(board [][]byte, r, c int) bool {
	if c == 9 {
		r++
		c = 0
	}
	if r == 9 {
		return true
	}

	if board[r][c] == '.' {
		for i := 1; i < 10; i++ {
			if row[r][i] == 1 {
				continue
			} else if col[c][i] == 1 {
				continue
			} else if grid[r/3][c/3][i] == 1 {
				continue
			}

			row[r][i] = 1
			col[c][i] = 1
			grid[r/3][c/3][i] = 1
			board[r][c] = byte(i) + '0'

			// search deeper
			if _0037_helper(board, r, c+1) {
				return true
			} else {
				row[r][i] = 0
				col[c][i] = 0
				grid[r/3][c/3][i] = 0
				board[r][c] = '.'
				continue
			}
		}
		return false
	} else {
		return _0037_helper(board, r, c+1)
	}
}

func main() {
	board := [][]byte{
		{'5', '3', '.', '.', '7', '.', '.', '.', '.'},
		{'6', '.', '.', '1', '9', '5', '.', '.', '.'},
		{'.', '9', '8', '.', '.', '.', '.', '6', '.'},
		{'8', '.', '.', '.', '6', '.', '.', '.', '3'},
		{'4', '.', '.', '8', '.', '3', '.', '.', '1'},
		{'7', '.', '.', '.', '2', '.', '.', '.', '6'},
		{'.', '6', '.', '.', '.', '.', '2', '8', '.'},
		{'.', '.', '.', '4', '1', '9', '.', '.', '5'},
		{'.', '.', '.', '.', '8', '.', '.', '7', '9'},
	}
	solveSudoku(board)
	return
}

File: solution/0037____Sudoku_Solver.cpp
#include "util\frequently_included.h"

inline int get_sub_index(int row, int col){
        return (row / 3) * 3 + col / 3;
}

class Solution {
public:
        void solveSudoku(vector<vector<char>>& board) {
                _1(board);
        }
private:
        void _1(vector<vector<char>>& board){
                vector<int> sub(9, 0);
                for(int i = 0; i < 9; ++i)
                        for(int ii = 0; ii < 9; ++ii)
                                if(board[i][ii] != '.'){
                                        char temp = board[i][ii] - '0';
                                        sub[get_sub_index(i, ii)] |= 1 << temp;
                                }
                _1_dfs(board, sub, 0, 0);
        }
        bool _1_dfs(vector<vector<char>>& board, vector<int>& sub, int i, int ii){
                if(ii == 9){
                        i++;
                        ii = 0;
                }
                if(i == 9)
                        return true;
                if(board[i][ii] == '.'){
                        for(int mask = 1; mask <= 9; ++mask){
                                int temp = 1 << mask;
                                if(!(temp & row[i]) && !(temp & col[i]) && !(temp & sub[get_sub_index(i, ii)])){
                                        row[i] |= temp;
                                        col[ii] |= temp;
                                        sub[get_sub_index(i, ii)] |= temp;
                                        board[i][ii] = '0' + mask;
                                        if(_1_dfs(board, row, col, sub, i, ii + 1))
                                                return true;
                                        row[i] ^= temp;
                                        col[ii] ^= temp;
                                        sub[get_sub_index(i, ii)] ^= temp;
                                        board[i][ii] = '.';
                                }
                        }
                        return false;
                }else{
                        return _1_dfs(board, row, col, sub, i, ii + 1);
                }
        }
};

int main(){
        vector<vector<char>> board = {
                {'5','3','.','.','7','.','.','.','.'},
                {'6','.','.','1','9','5','.','.','.'},
                {'.','9','8','.','.','.','.','6','.'},
                {'8','.','.','.','6','.','.','.','3'},
                {'4','.','.','8','.','3','.','.','1'},
                {'7','.','.','.','2','.','.','.','6'},
                {'.','6','.','.','.','.','2','8','.'},
                {'.','.','.','4','1','9','.','.','5'},
                {'.','.','.','.','8','.','.','7','9'}
        };
        Solution sln;
        sln.solveSudoku(board);
        return 0;
}

